src/LazyArtifacts.jl

```diff
@@ -9,7 +9,72 @@ using Artifacts: Artifacts,
 export artifact_exists, artifact_path, artifact_meta, artifact_hash,
        select_downloadable_artifacts, find_artifacts_toml, @artifact_str
 
-# define a function for satisfying lazy Artifact downloads
-using Pkg.Artifacts: ensure_artifact_installed
+using Base.BinaryPlatforms: AbstractPlatform, HostPlatform
+using Base: SHA1
+
+# We are mimicking Pkg.Artifacts.ensure_artifact_installed here so that we can
+# check if the artifact is already installed before loading Pkg.
+# Then if we need to load Pkg we do it in a subprocess to avoid precompilation
+# complexity.
+
+"""
+       ensure_artifact_installed(name::String, artifacts_toml::String;
+                                   platform::AbstractPlatform = HostPlatform(),
+                                   pkg_uuid::Union{Base.UUID,Nothing}=nothing,
+                                   verbose::Bool = false,
+                                   quiet_download::Bool = false,
+                                   io::IO=stderr)
+
+Ensures an artifact is installed, downloading it via the download information stored in
+`artifacts_toml` if necessary.  Throws an error if unable to install.
+"""
+function ensure_artifact_installed(name::String, artifacts_toml::String;
+    platform::AbstractPlatform=HostPlatform(),
+    pkg_uuid::Union{Base.UUID,Nothing}=nothing,
+    verbose::Bool=false,
+    quiet_download::Bool=false,
+    io::IO=stderr)
+    meta = artifact_meta(name, artifacts_toml; pkg_uuid=pkg_uuid, platform=platform)
+    if meta === nothing
+        error("Cannot locate artifact '$(name)' in '$(artifacts_toml)'")
+    end
+
+    return ensure_artifact_installed(name, meta, artifacts_toml;
+        platform, verbose, quiet_download, io)
+end
+
+function ensure_artifact_installed(name::String, meta::Dict, artifacts_toml::String;
+    platform::AbstractPlatform=HostPlatform(),
+    verbose::Bool=false,
+    quiet_download::Bool=false,
+    io::IO=stderr)
+
+    hash = SHA1(meta["git-tree-sha1"])
+    if !artifact_exists(hash)
+        # loading Pkg is a bit slow, so we only do it if we need to
+        # and do it in a subprocess to avoid precompilation complexity
+        code = """
+            Pkg = Base.require_stdlib(Base.PkgId(Base.UUID("44cfe95a-1eb2-52ea-b672-e2afdf69b78f"), "Pkg"));
+            Pkg.Artifacts.try_artifact_download_sources(
+                $(repr(name)),
+                Base.$(repr(hash)),
+                $(repr(meta)),
+                $(repr(artifacts_toml));
+                platform = Base.BinaryPlatforms.$(repr(platform)),
+                verbose = $(repr(verbose)),
+                quiet_download = $(repr(quiet_download)),
+                io = stderr
+            )
```

Comment:
Hmm, unfortunately running a sub-process with dynamic code like this is very incompatible with `--trim`

Reply:
See https://github.com/JuliaPackaging/LazyArtifacts.jl/pull/1#discussion_r1752202752 for some discussion

Reply:
And https://github.com/JuliaLang/julia/issues/55725#issuecomment-2340884118

Reply:
Is it possible to do this only during pre-compilation? That behavior on its own is fine, but doing it at runtime is problematic for `--trim`

Reply:
> Is it possible to do this only during pre-compilation?

If we revert https://github.com/JuliaPackaging/JLLWrappers.jl/pull/66 yes :stuck_out_tongue:

Reply:
SGTM. Pushed

Reply:
Yeah, that JLLWrappers fix is helpful for this. We should rarely see lazy artifact installs during precompilation now.

Reply:
@topolarity's point is that he _wants_ the download during precompilation, because if you try to compile Julia ahead-of-time, you'll run into these code paths at runtime, which is not good. For those usecases though, I feel we need a more comprehensive solution for ensuring that things are installed and ready before things get run.  Like, lazy artifacts in general are just completely incompatible with the philosophy of AOT compilation.  That's why e.g. `PackageCompiler` downloads all lazy artifacts first.

Reply:
I don't really agree that lazy artifacts are incompatible with AOT compilation - but they are definitely incompatible with the idea of providing a self-contained executable/package

My request was for this to be branched on `Base.generating_output()` because the `true` branch is not inferable/compilable for `juliac`, but we have ways to let `juliac` know that only the `false` branch is ever taken at run-time.

In that case, you're still running an AOT-compiled executable and the download works - it just happens lazily.

Reply:
I think the question here is more about legality, rather than preference that something orchestrating `--trim` would have for artifact downloads a la PackageCompiler.

As it is now seems more generally legal.

Reply:
On second thought, I fear that `Base.require_stdlib` is still going to be effectively impossible for us to infer

We're not even allowed to load the stdlib for inference on its own functions, because that would break the same pre-compilation assumptions we're skirting in the other branch.

I'm a bit worried that this change just makes LazyArtifacts incompatible with `--trim` unfortunately

We might have to find a way to load Pkg eagerly for `--trim` alone

Reply:
I don't know if you can infer it directly, but you should be able to pre-emulate it, since it is just a shim for finding the right `.ji` file to load, and that can be pre-computed and run directly (esp. for --trim)

Reply:
Right but then I need to infer `Pkg.Artifacts.try_artifact_download_sources` and also ensure that its code is available in the loaded pkgimage (or our own sysimage)